### PR TITLE
Fix Issue 23705 - dmd: src/dmd/backend/cgcod.d:734: Assertion sz >= 0 failed

### PR DESCRIPTION
**Pull request by RazvanN7**
Targeting master as this manifests in non-release builds, although it might have negative effects on release builds also.

### CODE REVIEW

File: compiler/src/dmd/backend/cgcod.d [hunk 1]
```diff
@@ -725,12 +725,12 @@ tryagain:
 @trusted
 targ_size_t alignsection(targ_size_t base, uint alignment, int bias)
 {
-    assert(cast(int)base <= 0);
+    assert(cast(long)base <= 0);
```
Comment (RazvanN7) [inline compiler/src/dmd/backend/cgcod.d:728]: `targ_size_t` is an alias to `ulong`, therefore casting to `int` gets a part of the number.

**Review by dkorpel — APPROVED** [review]

**Comment by ibuclaw** [comment]
Isn't release logically deprecated? ICE is an ICE is an ICE. You should never be able to crash the compiler whatever you're doing.

**Comment by RazvanN7** [comment]
Fair enough. Targeting stable now.

**Review by thewilsonator — APPROVED** [review]

File: compiler/test/compilable/test23705.d [hunk 2]
```diff
@@ -0,0 +1,8 @@
+// https://issues.dlang.org/show_bug.cgi?id=23705
+
+// DISABLED: win32
+
+void main ()
+{
+    ubyte [0x7fff_fffe] x;
```
Comment (ibuclaw) [inline compiler/test/compilable/test23705.d:7]: This test fails on many platforms - especially i686 and other 32-bit architectures - as there is no space for the frame pointer.

Comment (dkorpel) [reply compiler/test/compilable/test23705.d:7]: I don't think so, but perhaps `static if (size_t.sizeof > 4)` could be added.

Comment (ibuclaw) [reply compiler/test/compilable/test23705.d:7]: FAOD, GCC enforces that the maximum size for local variables to occupy the stack to be:
```
ubyte [(1 << (size_t.sizeof * 8 - 1)) - 64 * size_t.sizeof] x;
```
That is per function, not per variable.

**Pull request merged**